WebClient: Broken order of Netty writes

[//]: # "Please remove these comments"
[//]: # "Provide a general summary of the issue in the Title above"

## Environment Details
* Helidon Version: 2.4.1-SNAPSHOT
* Helidon SE
* JDK version:
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.0.2 (build 11.0.10+8-jvmci-21.0-b06, mixed mode, sharing)
* OS:
Distributor ID:	Ubuntu
Description:	Ubuntu 21.10
Release:	21.10
Codename:	impish
----------

## Problem Description
WebClient break Netty order of writes when consuming stream powered by multiple threads.

## Steps to reproduce
```java
        final ExecutorService exec = Executors.newSingleThreadExecutor();
        WebServer server = WebServer.builder()
                    .routing(Routing.builder()
                            .post((req, res) -> req.content()
                                    .onComplete(res::send)
                                    .onError(res::send)
                                    .ignoreElements())
                            .build())
                    .host("localhost")
                    .port(0)
                    .build()
                    .start()
                    .await(TIME_OUT);

            WebClient.builder()
                    .baseUri("http://localhost:" + server.port())
                    .connectTimeout(TIME_OUT.toMillis(), TimeUnit.MILLISECONDS)
                    .build()
                    .post()
                    .submit(Multi.just("1", "2", "3")
                            .map(String::valueOf)
                            .map(String::getBytes)
                            // Condition: initiate write from different thread than event loop
                            .observeOn(exec)
                            .map(n -> DataChunk.create(ByteBuffer.wrap(n)))
                            // Condition: mix-up upstream threads, flatMap has a prefetch with cache
                            // eg. some onNexts are going to be served by requesting thread
                            .flatMap(Single::just)
                    )
                    .await(TIME_OUT);
```
Ends with:
```
java.util.concurrent.CompletionException: java.lang.IllegalStateException: (client reqID: 1) Failure when sending a content!

	at io.helidon.common.reactive@2.4.1-SNAPSHOT/io.helidon.common.reactive.Awaitable.await(Awaitable.java:90)
	at io.helidon.webserver@2.4.1-SNAPSHOT/io.helidon.webserver.OrderOfWritesTest.threadMixUp2(OrderOfWritesTest.java:137)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
...

Caused by: java.lang.IllegalStateException: (client reqID: 1) Failure when sending a content!
	at io.helidon.webclient.RequestContentSubscriber.lambda$completeOnFailureListener$8(RequestContentSubscriber.java:185)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:578)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:571)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:550)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:491)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:616)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:609)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
	at io.netty.common@4.1.63.Final/io.netty.util.internal.PromiseNotificationUtil.tryFailure(PromiseNotificationUtil.java:64)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.notifyOutboundHandlerException(AbstractChannelHandlerContext.java:814)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:719)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702)
	at io.netty.handler@4.1.63.Final/io.netty.handler.timeout.IdleStateHandler.write(IdleStateHandler.java:302)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.transport@4.1.63.Final/io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.common@4.1.63.Final/io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.common@4.1.63.Final/io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.netty.handler.codec.EncoderException: java.lang.IllegalStateException: unexpected message type: DefaultHttpContent, state: 0
	at io.netty.codec@4.1.63.Final/io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:104)
	at io.netty.transport@4.1.63.Final/io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
	at io.netty.transport@4.1.63.Final/io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	... 13 more
Caused by: java.lang.IllegalStateException: unexpected message type: DefaultHttpContent, state: 0
	at io.netty.codec.http@4.1.63.Final/io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:124)
	at io.netty.codec.http@4.1.63.Final/io.netty.handler.codec.http.HttpClientCodec$Encoder.encode(HttpClientCodec.java:189)
	at io.netty.codec@4.1.63.Final/io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:89)
	... 15 more
```

--- a/webclient/webclient/pom.xml
+++ b/webclient/webclient/pom.xml
@@ -93,4 +93,31 @@
             <scope>test</scope>
         </dependency>
     </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>default-compile</id>
+                        <configuration>
+                            <failOnWarning>true</failOnWarning>
+                        </configuration>
+                    </execution>
+                    <execution>
+                        <id>default-testCompile</id>
+                        <configuration>
+                            <failOnWarning>true</failOnWarning>
+                            <compilerArgs>
+                                <arg>--add-modules</arg>
+                                <arg>jdk.httpserver</arg>
+                            </compilerArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
 </project>

--- a/webclient/webclient/src/main/java/io/helidon/webclient/NettyChannel.java
+++ b/webclient/webclient/src/main/java/io/helidon/webclient/NettyChannel.java
@@ -1,0 +1,129 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.webclient;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+
+import static io.helidon.webclient.WebClientRequestBuilderImpl.RECEIVED;
+import static io.helidon.webclient.WebClientRequestBuilderImpl.REQUEST;
+
+/**
+ * Wrapper for {@link io.netty.channel.ChannelHandlerContext} guarding all writes to be made exclusively from event loop thread.
+ *
+ * <ul>
+ * <b>Netty can provide "ordering" in the one of the following situations exclusively:</b>
+ *  <ul>
+ *      <li>You are doing all writes from the EventLoop thread.</li>
+ *      <li>You are doing no writes from the EventLoop thread (i.e. all writes are being done in other thread(s)).</li>
+ *  </ul>
+ * </ul>
+ *
+ * @see <a href="https://github.com/netty/netty/issues/3887#issuecomment-112540327"
+ * >https://github.com/netty/netty/issues/3887#issuecomment-112540327</a>
+ */
+class NettyChannel {
+
+    private final Channel channel;
+    private CompletionStage<ChannelFuture> writeFuture = CompletableFuture.completedFuture(null);
+
+    NettyChannel(Channel channel) {
+        this.channel = channel;
+    }
+
+    /**
+     * Request to flush all pending messages via this ChannelOutboundInvoker from Netty's event loop thread.
+     */
+    void flush() {
+        writeFuture = writeFuture.thenApply(f -> {
+            if (channel.eventLoop().inEventLoop()) {
+                channel.flush();
+            } else {
+                channel.eventLoop().execute(channel::flush);
+            }
+            return f;
+        });
+    }
+
+    /**
+     * Request to write a message via this ChannelHandlerContext through the ChannelPipeline from Netty's event loop thread.
+     *
+     * @param flush flush immediately
+     * @param msg   message to write
+     * @param listeners function for accessing channel future of the netty write
+     * @return CompletionStage completed when request is made from event loop thread, containing future of actual write
+     */
+    void write(boolean flush, Object msg, Function<ChannelFuture, ChannelFuture> listeners) {
+        // Ordered writes
+        writeFuture = writeFuture.thenCompose(f -> writeInt(flush, msg, listeners));
+    }
+
+    /**
+     * Request to write a message via this ChannelHandlerContext through the ChannelPipeline from Netty's event loop thread.
+     *
+     * @param flush flush immediately
+     * @param msg   message to write
+     * @return CompletionStage completed when request is made from event loop thread, containing future of actual write
+     */
+    void write(boolean flush, Object msg) {
+        // Ordered writes
+        writeFuture = writeFuture.thenCompose(f -> writeInt(flush, msg, Function.identity()));
+    }
+
+    private CompletionStage<ChannelFuture> writeInt(boolean flush,
+                                                    Object msg,
+                                                    Function<ChannelFuture, ChannelFuture> listeners) {
+        CompletableFuture<ChannelFuture> channelFuture = new CompletableFuture<>();
+
+        if (channel.eventLoop().inEventLoop()) {
+            // Fast path for items emitted by event loop thread
+            if (flush) {
+                channelFuture.complete(listeners.apply(channel.writeAndFlush(msg)));
+            } else {
+                channelFuture.complete(listeners.apply(channel.write(msg)));
+            }
+        } else {
+            if (flush) {
+                channel.eventLoop().execute(() -> channelFuture.complete(listeners.apply(channel.writeAndFlush(msg))));
+            } else {
+                channel.eventLoop().execute(() -> channelFuture.complete(listeners.apply(channel.write(msg))));
+            }
+        }
+
+        return channelFuture;
+    }
+
+    WebClientServiceRequest serviceRequest() {
+        return channel.attr(REQUEST).get().configuration().clientServiceRequest();
+    }
+
+    boolean isConnectionReset() {
+        return channel.attr(RECEIVED).get().isDone() || !channel.isActive();
+    }
+
+    @Override
+    public String toString() {
+        return "NettyChannel{"
+                + "ctx=" + channel.toString()
+                + '}';
+    }
+}

--- a/webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java
+++ b/webclient/webclient/src/main/java/io/helidon/webclient/RequestContentSubscriber.java
@@ -37,8 +37,6 @@ import io.netty.handler.codec.http.LastHttpContent;
 import io.netty.util.concurrent.Future;
 import io.netty.util.concurrent.GenericFutureListener;
 
-import static io.helidon.webclient.WebClientRequestBuilderImpl.RECEIVED;
-import static io.helidon.webclient.WebClientRequestBuilderImpl.REQUEST;
 import static io.helidon.webclient.WebClientRequestBuilderImpl.REQUEST_ID;
 
 /**
@@ -53,7 +51,7 @@ class RequestContentSubscriber implements Flow.Subscriber<DataChunk> {
     private final CompletableFuture<WebClientResponse> responseFuture;
     private final CompletableFuture<WebClientServiceRequest> sent;
     private final DefaultHttpRequest request;
-    private final Channel channel;
+    private final NettyChannel channel;
     private final long requestId;
     private final boolean allowChunkedEncoding;
 
@@ -67,7 +65,7 @@ class RequestContentSubscriber implements Flow.Subscriber<DataChunk> {
                              CompletableFuture<WebClientServiceRequest> sent,
                              boolean allowChunkedEncoding) {
         this.request = request;
-        this.channel = channel;
+        this.channel = new NettyChannel(channel);
         this.responseFuture = responseFuture;
         this.sent = sent;
         this.requestId = channel.attr(REQUEST_ID).get();
@@ -113,7 +111,7 @@ class RequestContentSubscriber implements Flow.Subscriber<DataChunk> {
                                                          + Http.Header.CONNECTION + ": close, has to be set.");
                 }
             }
-            channel.writeAndFlush(request);
+            channel.write(true, request);
             sendData(firstDataChunk);
             firstDataChunk = null;
 
@@ -141,15 +139,14 @@ class RequestContentSubscriber implements Flow.Subscriber<DataChunk> {
             } else if (HttpUtil.getContentLength(request) == 0 && firstDataChunk != null) {
                 HttpUtil.setContentLength(request, firstDataChunk.remaining());
             }
-            channel.writeAndFlush(request);
+            channel.write(true, request);
             if (firstDataChunk != null) {
                 sendData(firstDataChunk);
             }
         }
-        WebClientRequestImpl clientRequest = channel.attr(REQUEST).get();
-        WebClientServiceRequest serviceRequest = clientRequest.configuration().clientServiceRequest();
+        WebClientServiceRequest serviceRequest = channel.serviceRequest();
         LOGGER.finest(() -> "(client reqID: " + requestId + ") Sending last http content");
-        channel.writeAndFlush(LAST_HTTP_CONTENT)
+        channel.write(true, LAST_HTTP_CONTENT, f -> f
                 .addListener(completeOnFailureListener("(client reqID: " + requestId + ") "
                                                                + "An exception occurred when writing last http content."))
                 .addListener(ChannelFutureListener.CLOSE_ON_FAILURE)
@@ -158,27 +155,27 @@ class RequestContentSubscriber implements Flow.Subscriber<DataChunk> {
                         sent.complete(serviceRequest);
                         LOGGER.finest(() -> "(client reqID: " + requestId + ") Request sent");
                     }
-                });
+                }));
     }
 
     private void sendData(DataChunk data) {
         LOGGER.finest(() -> "(client reqID: " + requestId + ") Sending data chunk");
         DefaultHttpContent httpContent = new DefaultHttpContent(Unpooled.wrappedBuffer(data.data()));
-        channel.writeAndFlush(httpContent)
+        channel.write(true, httpContent, f -> f
                 .addListener(future -> {
                     data.release();
                     subscription.request(1);
                     LOGGER.finest(() -> "(client reqID: " + requestId + ") Data chunk sent with result: " + future.isSuccess());
                 })
                 .addListener(completeOnFailureListener("(client reqID: " + requestId + ") Failure when sending a content!"))
-                .addListener(ChannelFutureListener.CLOSE_ON_FAILURE);
+                .addListener(ChannelFutureListener.CLOSE_ON_FAILURE));
     }
 
     private GenericFutureListener<Future<? super Void>> completeOnFailureListener(String message) {
         return future -> {
             if (!future.isSuccess()) {
                 Throwable cause = future.cause();
-                if (channel.attr(RECEIVED).get().isDone() || !channel.isActive()) {
+                if (channel.isConnectionReset()) {
                     completeRequestFuture(new IllegalStateException("(client reqID: " + requestId + ") "
                                                                             + "Connection reset by the host", cause));
                 } else {

--- a/webclient/webclient/src/test/java/io/helidon/webclient/OrderOfWritesTest.java
+++ b/webclient/webclient/src/test/java/io/helidon/webclient/OrderOfWritesTest.java
@@ -1,0 +1,91 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.webclient;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import io.helidon.common.LogConfig;
+import io.helidon.common.http.DataChunk;
+import io.helidon.common.reactive.Multi;
+import io.helidon.common.reactive.Single;
+
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.sun.net.httpserver.HttpServer;
+
+public class OrderOfWritesTest {
+    private static final Duration TIME_OUT = Duration.ofSeconds(5);
+
+    @Test
+    void threadMixUp() throws IOException, InterruptedException {
+        LogConfig.configureRuntime();
+        ExecutorService exec = null;
+        HttpServer server = null;
+        CompletableFuture<String> resultFuture = new CompletableFuture<>();
+        try {
+            exec = Executors.newFixedThreadPool(3);
+            server = HttpServer.create(new InetSocketAddress(0), 0);
+            server.createContext("/", res -> {
+                resultFuture.complete(new String(res.getRequestBody().readAllBytes()));
+                res.sendResponseHeaders(200, 0);
+            });
+            server.setExecutor(exec);
+            server.start();
+
+            WebClient.builder()
+                    .baseUri("http://localhost:" + server.getAddress().getPort())
+                    .connectTimeout(TIME_OUT.toMillis(), TimeUnit.MILLISECONDS)
+                    .build()
+                    .post()
+                    .submit(Multi.just("1", "2", "3")
+                            .map(String::valueOf)
+                            .map(String::getBytes)
+                            // Condition: initiate write from different thread than event loop
+                            .observeOn(exec)
+                            .map(n -> DataChunk.create(ByteBuffer.wrap(n)))
+                            // Condition: mix-up upstream threads, flatMap has a prefetch with cache
+                            // eg. some onNexts are going to be served by requesting thread
+                            .flatMap(Single::just, 2, true, 2)
+                    )
+                    .await(TIME_OUT);
+
+            String result = Single.create(resultFuture).await(TIME_OUT);
+
+            assertThat(result, Matchers.equalTo("123"));
+        } finally {
+            if (server != null) {
+                server.stop(0);
+            }
+            if (exec != null) {
+                exec.shutdownNow();
+                assertTrue(exec.awaitTermination(TIME_OUT.toMillis(), TimeUnit.MILLISECONDS));
+            }
+        }
+    }
+}

